substrate/frame/migrations/src/tests.rs

```diff
@@ -0,0 +1,369 @@
+
+#[test]
+#[docify::export]
+fn simple_works() {
+	use Event::*;
+	test_closure(|| {
+		// Add three migrations, each taking one block longer than the last.
+		MigrationsStorage::set(vec![(SucceedAfter, 0), (SucceedAfter, 1), (SucceedAfter, 2)]);
+
+		System::set_block_number(1);
+		Migrations::on_runtime_upgrade();
+		run_to_block(10);
+
+		// Check that the executed migrations are recorded in `Historical`.
+		assert_eq!(
+			historic(),
+			vec![
+				mocked_id(SucceedAfter, 0),
+				mocked_id(SucceedAfter, 1),
+				mocked_id(SucceedAfter, 2),
+			]
+		);
+		// Check that we got all events.
+		assert_events(vec![
+			UpgradeStarted { migrations: 3 },
+			MigrationCompleted { index: 0, blocks: 1 },
+			MigrationAdvanced { index: 1, blocks: 0 },
+			MigrationCompleted { index: 1, blocks: 1 },
+			MigrationAdvanced { index: 2, blocks: 0 },
+			MigrationAdvanced { index: 2, blocks: 1 },
+			MigrationCompleted { index: 2, blocks: 2 },
+			UpgradeCompleted,
+		]);
+	});
+}
+
+#[test]
+fn basic_works() {
+	test_closure(|| {
+		MigrationsStorage::set(vec![(SucceedAfter, 0), (SucceedAfter, 1), (SucceedAfter, 2)]);
+
+		System::set_block_number(1);
+		Migrations::on_runtime_upgrade();
+		run_to_block(10);
+
+		// Check that the executed migrations are recorded into `Historical`.
+		assert_eq!(
+			historic(),
+			vec![
+				mocked_id(SucceedAfter, 0),
+				mocked_id(SucceedAfter, 1),
+				mocked_id(SucceedAfter, 2),
+			]
+		);
+		// Check that we got all events.
+		assert_events(vec![
+			Event::UpgradeStarted { migrations: 3 },
+			Event::MigrationCompleted { index: 0, blocks: 1 },
+			Event::MigrationAdvanced { index: 1, blocks: 0 },
+			Event::MigrationCompleted { index: 1, blocks: 1 },
+			Event::MigrationAdvanced { index: 2, blocks: 0 },
+			Event::MigrationAdvanced { index: 2, blocks: 1 },
+			Event::MigrationCompleted { index: 2, blocks: 2 },
+			Event::UpgradeCompleted,
+		]);
+	});
+}
+
+#[test]
+fn failing_migration_keep_stuck_the_chain() {
+	test_closure(|| {
+		FailedUpgradeResponse::set(FailedUpgradeHandling::KeepStuck);
+		MigrationsStorage::set(vec![(FailAfter, 2)]);
+
+		System::set_block_number(1);
+		Migrations::on_runtime_upgrade();
+		run_to_block(10);
+
+		// Failed migrations are not recorded in `Historical`.
+		assert!(historic().is_empty());
+		// Check that we got all events.
+		assert_events(vec![
+			Event::UpgradeStarted { migrations: 1 },
+			Event::MigrationAdvanced { index: 0, blocks: 1 },
+			Event::MigrationAdvanced { index: 0, blocks: 2 },
+			Event::MigrationFailed { index: 0, blocks: 3 },
+			Event::UpgradeFailed,
+		]);
+
+		// Check that the handler was called correctly.
+		assert_eq!(UpgradesStarted::take(), 1);
+		assert_eq!(UpgradesCompleted::take(), 0);
+		assert_eq!(UpgradesFailed::take(), vec![Some(0)]);
+
+		assert_eq!(Cursor::<T>::get(), Some(MigrationCursor::Stuck), "Must stuck the chain");
+	});
+}
+
+#[test]
+fn failing_migration_force_unstuck_the_chain() {
```

Comment:
maybe this is better? 
```suggestion
fn failing_migration_force_unstuck_works() {
```